File: server/screenshot.lua
local config = require 'config.server'

local pending = {}
local counter = 0

AddEventHandler('qbx_properties:internal:shotResult', function(id, result)
    local p = pending[id]
    if not p then return end
    pending[id] = nil
    p:resolve(result)
end)

local function requestCapture(event, ...)
    counter += 1
    local id = counter
    local p = promise.new()
    pending[id] = p
    TriggerEvent(event, id, ...)
    SetTimeout(15000, function()
        if pending[id] then
            pending[id] = nil
            p:resolve({ ok = false, error = 'timeout' })
        end
    end)
    return Citizen.Await(p)
end

local imageProviders = {
    qbox = {
        url = 'https://api.qbox.re/v1/file', field = 'file', responsePath = 'data.url',
        storagePath = 'data.path', deleteUrl = 'https://api.qbox.re/v1/file/%s',
    },
    fivemanage = {
        url = 'https://api.fivemanage.com/api/v3/file', field = 'file', responsePath = 'data.url',
        storagePath = 'data.id', deleteUrl = 'https://api.fivemanage.com/api/v3/file/%s',
    },
    fivemerr = { url = 'https://api.fivemerr.com/v1/media/images', field = 'file', responsePath = 'url' },
}

local function imageProvider()
    local cfg = config.imageUpload
    if not cfg or not cfg.apiKey or cfg.apiKey == '' then return end
    local provider = cfg.provider == 'custom' and cfg.custom or imageProviders[cfg.provider]
    if not provider or not provider.url or provider.url == '' then return end
    return {
        url = provider.url,
        field = provider.field or 'file',
        responsePath = provider.responsePath or 'url',
        storagePath = provider.storagePath,
        deleteUrl = provider.deleteUrl,
        apiKey = cfg.apiKey,
    }
end

local function isAdmin(source)
    return exports.qbx_core:HasPermission(source, 'admin')
end

lib.callback.register('qbx_properties:callback:furnitureShotWrite', function(source, filename, base64)
    if not isAdmin(source) then return { ok = false, error = 'no permission' } end
    local result = requestCapture('qbx_properties:internal:shotWrite', tostring(filename), base64)
    if result and result.ok and result.file and config.imageUpload and config.imageUpload.autoUpload then
        local provider = imageProvider()
        if provider then
            TriggerEvent('qbx_properties:internal:cdnQueue', result.file, provider)
        end
    end
    return result
end)

lib.callback.register('qbx_properties:callback:furnitureShotList', function(source)
    if not isAdmin(source) then return { ok = false } end
    return requestCapture('qbx_properties:internal:shotList')
end)

lib.callback.register('qbx_properties:callback:furnitureShotTuning', function(source)
    if not isAdmin(source) then return { ok = false } end
    return requestCapture('qbx_properties:internal:getTuning')
end)

lib.callback.register('qbx_properties:callback:furnitureShotSaveTuning', function(source, tuning)
    if not isAdmin(source) or type(tuning) ~= 'table' then return { ok = false } end
    return requestCapture('qbx_properties:internal:saveTuning', tuning)
end)

lib.callback.register('qbx_properties:callback:furnitureCdnMap', function()
    local result = requestCapture('qbx_properties:internal:cdnMap')
    return result and result.ok and result.map or {}
end)

local cdnSyncing = false

local function cdnSync(source, mode)
    local provider = imageProvider()
    if not provider then
        exports.qbx_core:Notify(source, 'No image upload provider configured, set imageUpload in config/server.lua.', 'error')
        return
    end
    if cdnSyncing then
        exports.qbx_core:Notify(source, 'An upload is already running.', 'error')
        return
    end

    local listed = requestCapture('qbx_properties:internal:shotList')
    if not listed or not listed.ok then
        exports.qbx_core:Notify(source, 'Could not list the screenshots.', 'error')
        return
    end
    local mapped = requestCapture('qbx_properties:internal:cdnMap')
    local map = mapped and mapped.ok and mapped.map or {}

    local missing = {}
    for i = 1, #listed.files do
        local file = listed.files[i]
        if mode == 'all' or not map[file] then missing[#missing + 1] = file end
    end
    if #missing == 0 then
        exports.qbx_core:Notify(source, 'Every screenshot is already on the CDN.', 'success')
        return
    end

    cdnSyncing = true
    exports.qbx_core:Notify(source, ('Uploading %d image(s) in the background.'):format(#missing), 'info')
    CreateThread(function()
        local failed = 0
        for i = 1, #missing do
            local result = requestCapture('qbx_properties:internal:cdnUpload', missing[i], provider)
            if not result or not result.ok then
                failed += 1
                lib.print.warn(('cdn upload failed for %s: %s'):format(missing[i], result and result.error or 'timeout'))
            end
            if i % 100 == 0 then
                lib.print.info(('cdn sync: %d/%d uploaded'):format(i, #missing))
            end
        end
        cdnSyncing = false
        lib.print.info(('cdn sync finished: %d uploaded, %d failed'):format(#missing - failed, failed))
        if GetPlayerName(source) then
            exports.qbx_core:Notify(source, ('CDN upload finished: %d uploaded, %d failed.'):format(#missing - failed, failed), failed > 0 and 'error' or 'success', 8000)
        end
    end)
end

lib.addCommand('screenshotfurniture', {
    help = 'Regenerate the furniture catalog images',
    params = {
        { name = 'mode', help = "'overwrite' reshoots existing images, 'manual' starts paused with full control, 'upload' pushes missing images to the CDN, 'uploadall' re-uploads everything", type = 'string', optional = true },
    },
    restricted = 'group.admin',
}, function(source, args)
    local mode = args.mode
    if mode == 'upload' or mode == 'uploadall' then
        cdnSync(source, mode == 'uploadall' and 'all' or 'missing')
        return
    end
    TriggerClientEvent('qbx_properties:client:screenshotFurniture', source, mode)
end)

File: server/screenshot.js
const fs = require('fs');
const path = require('path');

const RESOURCE_NAME = GetCurrentResourceName();
const SCREENSHOTS_DIR = path.join(GetResourcePath(RESOURCE_NAME), 'screenshots');

if (!fs.existsSync(SCREENSHOTS_DIR)) fs.mkdirSync(SCREENSHOTS_DIR, { recursive: true });

function reply(id, result) {
  emit('qbx_properties:internal:shotResult', id, result);
}

on('qbx_properties:internal:shotWrite', (id, filename, base64) => {
  try {
    if (typeof filename !== 'string' || !/^[\w-]+$/.test(filename)) {
      return reply(id, { ok: false, error: 'unsafe filename' });
    }
    if (typeof base64 !== 'string' || base64.length > 2_000_000) {
      return reply(id, { ok: false, error: 'bad payload' });
    }
    const raw = base64.replace(/^data:image\/\w+;base64,/, '');
    const buffer = Buffer.from(raw, 'base64');
    if (buffer.length < 16 || buffer.toString('ascii', 0, 4) !== 'RIFF'
      || buffer.toString('ascii', 8, 12) !== 'WEBP') {
      return reply(id, { ok: false, error: 'not a webp' });
    }
    fs.writeFileSync(path.join(SCREENSHOTS_DIR, `${filename}.webp`), buffer);
    reply(id, { ok: true, file: `${filename}.webp` });
  } catch (err) {
    reply(id, { ok: false, error: String(err) });
  }
});

on('qbx_properties:internal:shotList', (id) => {
  try {
    const files = fs.readdirSync(SCREENSHOTS_DIR).filter((file) => file.endsWith('.webp'));
    reply(id, { ok: true, files });
  } catch (err) {
    reply(id, { ok: false, error: String(err) });
  }
});

const TUNING_PATH = path.join(SCREENSHOTS_DIR, '_tuning.json');

on('qbx_properties:internal:getTuning', (id) => {
  try {
    const tuning = fs.existsSync(TUNING_PATH) ? JSON.parse(fs.readFileSync(TUNING_PATH, 'utf8')) : {};
    reply(id, { ok: true, tuning });
  } catch (err) {
    reply(id, { ok: true, tuning: {} });
  }
});

on('qbx_properties:internal:saveTuning', (id, tuning) => {
  try {
    fs.writeFileSync(TUNING_PATH, JSON.stringify(tuning ?? {}, null, 2));
    reply(id, { ok: true });
  } catch (err) {
    reply(id, { ok: false, error: String(err) });
  }
});

const CDN_PATH = path.join(SCREENSHOTS_DIR, '_cdn.json');

let cdnMap = {};
try {
  if (fs.existsSync(CDN_PATH)) cdnMap = JSON.parse(fs.readFileSync(CDN_PATH, 'utf8'));
} catch { cdnMap = {}; }
for (const [key, value] of Object.entries(cdnMap)) {
  if (typeof value === 'string') cdnMap[key] = { url: value };
}

function saveCdnMap() {
  fs.writeFileSync(CDN_PATH, JSON.stringify(cdnMap, null, 2));
}

function responseField(value, fieldPath) {
  for (const key of fieldPath.split('.')) {
    if (typeof value !== 'object' || value === null) return;
    value = value[key];
  }
  return typeof value === 'string' ? value : undefined;
}

async function deleteRemote(entry, provider) {
  if (!entry?.path || !provider.deleteUrl) return;
  try {
    const res = await fetch(provider.deleteUrl.replace('%s', encodeURIComponent(entry.path)), {
      method: 'DELETE',
      headers: { Authorization: provider.apiKey },
      signal: AbortSignal.timeout(15_000),
    });
    if (!res.ok) console.log(`^3[qbx_properties] cdn delete returned ${res.status} for ${entry.path}^0`);
  } catch (err) {
    console.log(`^3[qbx_properties] cdn delete failed for ${entry.path}: ${err}^0`);
  }
}

async function uploadFile(relPath, provider) {
  const filePath = path.join(SCREENSHOTS_DIR, relPath);
  if (!filePath.startsWith(SCREENSHOTS_DIR) || !fs.existsSync(filePath)) {
    return { ok: false, error: 'file not found' };
  }
  const form = new FormData();
  form.append(provider.field || 'file',
    new Blob([fs.readFileSync(filePath)], { type: 'image/webp' }),
    relPath.replace(/[\\/]/g, '_'));
  const res = await fetch(provider.url, {
    method: 'POST',
    headers: { Authorization: provider.apiKey },
    body: form,
    signal: AbortSignal.timeout(30_000),
  });
  if (!res.ok) return { ok: false, error: `http ${res.status}` };
  const json = await res.json().catch(() => null);
  const url = json && responseField(json, provider.responsePath || 'url');
  if (!url) return { ok: false, error: 'no url in response' };

  const previous = cdnMap[relPath];
  if (previous && previous.url !== url) await deleteRemote(previous, provider);

  const storagePath = provider.storagePath ? responseField(json, provider.storagePath) : undefined;
  cdnMap[relPath] = storagePath ? { url, path: storagePath } : { url };
  saveCdnMap();
  return { ok: true, url };
}

on('qbx_properties:internal:cdnUpload', (id, relPath, provider) => {
  uploadFile(String(relPath), provider)
    .then((result) => reply(id, result))
    .catch((err) => reply(id, { ok: false, error: String(err) }));
});

on('qbx_properties:internal:cdnMap', (id) => {
  const map = {};
  for (const [key, entry] of Object.entries(cdnMap)) map[key] = entry.url;
  reply(id, { ok: true, map });
});

const uploadQueue = [];
let uploading = false;

async function drainQueue() {
  if (uploading) return;
  uploading = true;
  while (uploadQueue.length > 0) {
    const { relPath, provider } = uploadQueue.shift();
    try {
      const result = await uploadFile(relPath, provider);
      if (!result.ok) console.log(`^3[qbx_properties] cdn upload failed for ${relPath}: ${result.error}^0`);
    } catch (err) {
      console.log(`^3[qbx_properties] cdn upload failed for ${relPath}: ${err}^0`);
    }
  }
  uploading = false;
}

on('qbx_properties:internal:cdnQueue', (relPath, provider) => {
  uploadQueue.push({ relPath: String(relPath), provider });
  drainQueue();
});
